app/vlselect/logsql/logsql.go

```diff
@@ -420,9 +428,18 @@ func ProcessLiveTailRequest(ctx context.Context, w http.ResponseWriter, r *http.
 		logger.Panicf("BUG: it is expected that http.ResponseWriter (%T) supports http.Flusher interface", w)
 	}
 	qOrig := q
+	windowLimitExceeded := false
 	for {
-		start := end - tailOffsetNsecs
-		end = time.Now().UnixNano()
+		start := end - tailOffsetStartNsecs
+		end = time.Now().UnixNano() - tailOffsetEndNsecs
+
+		// Limit the search window to prevent out of memory situations when we fall too far behind.
+		if end-start > tailMaxWindowNsecs {
```

Comment:
While I understand that live-tailing is primarily intended for consumption by humans, I'm particularly worried about a situation where a human is consuming the live tail log, then pausing the stream and leaving it paused for a long time. Once the stream is resumed, the next query iteration consumes so much memory that it may cause VictoriaLogs to run out of memory. I consider this a DoS vulnerability even if not intentionally triggered.

Reply:
@Fusl , thanks for the explanation! This sounds reasonable, but this protection isn't going to be added because of the following reason.

VictoriaLogs querying API provides countless ways to shoot yourself in the foot and execute a query, which will eat all the available CPU and RAM (aka "DoS vulnerabilities"). For example, an innocent-looking query `* | uniq_values(_time)` may eat all the available memory, since it tries fetching `_time` field across all the logs and then returning them as JSON array via [`uniq_values` stats function](https://docs.victoriametrics.com/victorialogs/logsql/#uniq_values-stats). If the number of logs stored at VictoriaLogs is big (e.g. billions), then this query will definitely consume almost all the available disk read IO, CPU and RAM during the execution. It may not lead to OOM crash, since VicitoriaLogs contains memory usage tracking mechanism, which should automatically cancel the query with the error if it consumes too much memory. But this mechanism is easy to bypass by executing many heavy queries at the same time. The main purpose of this mechanism is to prevent from OOM crashes when users execute some heavy ad-hoc queries. It doesn't prevent from DoS attacks.

The threat model for VictoriaLogs is the following: it is expected that it runs in a protected network without direct access from the Internet. It is expected that access to it is provided via auth proxy such as [vmauth](https://docs.victoriametrics.com/vmauth/). VictoriaLogs doesn't provide protection from heavy queries to the supported [HTTP querying endpoints](https://docs.victoriametrics.com/victorialogs/querying/#http-api), since this is impossible task in general case. It is better to clearly document this (TBD) instead of complicating the code with partially-working protection.

Reply:
The security threat model for VictoriaLogs is documented at https://docs.victoriametrics.com/victorialogs/#security